zerver/views/streams.py

```diff
@@ -470,17 +479,25 @@ def remove_subscriptions_backend(
     for stream_name in streams_raw:
         streams_as_dict.append({"name": stream_name.strip()})
 
-    streams, __ = list_to_streams(
-        streams_as_dict, user_profile, unsubscribing_others=removing_someone_else
-    )
-
+    only_removing_own_bot = True
     if principals:
-        people_to_unsub = {
-            principal_to_user_profile(user_profile, principal) for principal in principals
-        }
+        people_to_unsub = set()
+        for principal in principals:
+            people = principal_to_user_profile(user_profile, principal)
+            people_to_unsub.add(people)
+            if user_profile == people:
+                continue
+            if only_removing_own_bot:
+                only_removing_own_bot = check_if_only_removing_own_bot(user_profile, people)
     else:
         people_to_unsub = {user_profile}
 
+    streams, __ = list_to_streams(
+        streams_as_dict,
+        user_profile,
+        unsubscribing_others=removing_someone_else and not only_removing_own_bot,
+    )
```

Comment:
This construction feels hard to understand. I think it's reasonable to construct `people_to_unsub` before calling `list_to_streams`; but then we should remove `removing_someone_else = check_if_removing_someone_else(user_profile, principals)`, and instead just compute `unsubscribing_others` by looping through `people_to_unsub` and checking if each user is either the current user or a bot owned by it.